zaza/utilities/openstack.py

```diff
@@ -1269,3 +1285,179 @@ def get_overcloud_auth():
             'API_VERSION': 3,
         }
     return auth_settings
+
+
+def get_urllib_opener():
+    """Create a urllib opener taking into account proxy settings
+
+    Using urllib.request.urlopen will automatically handle proxies so none
+    of this function is needed except we are currently specifying proxies
+    via AMULET_HTTP_PROXY rather than http_proxy so a ProxyHandler is needed
+    explicitly stating the proxies.
+
+    :returns: An opener which opens URLs via BaseHandlers chained together
+    :rtype: urllib.request.OpenerDirector
+    """
+    http_proxy = os.getenv('AMULET_HTTP_PROXY')
+    logging.debug('AMULET_HTTP_PROXY: {}'.format(http_proxy))
+
+    if http_proxy:
+        handler = urllib.request.ProxyHandler({'http': http_proxy})
+    else:
+        handler = urllib.request.HTTPHandler()
+    return urllib.request.build_opener(handler)
+
+
+def find_cirros_image(arch):
+    """Return the url for the latest cirros image for the given architecture
+
+    :param arch: aarch64, arm, i386, x86_64 etc
+    :type arch: str
+    :returns: Unit matching given name
+    :rtype: juju.unit.Unit or None
+    """
+    opener = get_urllib_opener()
+    f = opener.open(CIRROS_RELEASE_URL)
+    version = f.read().strip().decode()
+    cirros_img = 'cirros-{}-{}-disk.img'.format(version, arch)
+    return '{}/{}/{}'.format(CIRROS_IMAGE_URL, version, cirros_img)
+
+
+def download_image(image_url, target_file):
+    """Download the image from the given url to the specified file
+
+    :param image_url: URL to download image from
+    :type image_url: str
+    :param target_file: Local file to savee image to
+    :type target_file: str
+    """
+    opener = get_urllib_opener()
+    urllib.request.install_opener(opener)
+    urllib.request.urlretrieve(image_url, target_file)
+
+
+@tenacity.retry(wait=tenacity.wait_exponential(multiplier=1, max=60),
+                reraise=True, stop=tenacity.stop_after_attempt(8))
+def resource_reaches_status(resource, resource_id,
```

Comment:
Personal pet peeve. Can we clean up stat/state/status?  All three are used in this function. I personally preferer spelled out status or state.

Reply:
Done